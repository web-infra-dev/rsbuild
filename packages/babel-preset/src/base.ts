import { BabelOptions } from './types';

export type BasePresetOptions = {
  presetTypeScript?: Record<string, unknown> | false;
};

export const generateBaseConfig = (
  options: BasePresetOptions = {},
): BabelOptions => {
  const config: BabelOptions = {
    presets: [],
    plugins: [],
  };
  const { presetTypeScript = {} } = options;

  if (presetTypeScript) {
    config.presets?.push([
      require.resolve('@babel/preset-typescript'),
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        // aligns Babel's behavior with TypeScript's default behavior.
        // https://babeljs.io/docs/en/babel-preset-typescript#optimizeconstenums
        optimizeConstEnums: true,
        ...presetTypeScript,
      },
    ]);
  }

  return config;
};
