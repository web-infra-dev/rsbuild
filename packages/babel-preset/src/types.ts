import type {
  PresetEnvOptions,
  PresetEnvTargets,
  PresetEnvBuiltIns,
} from '@rsbuild/shared';

export type { TransformOptions as BabelConfig } from '@babel/core';

type DecoratorsVersion =
  | '2023-05'
  | '2023-01'
  | '2022-03'
  | '2021-12'
  | '2018-09'
  | 'legacy';

export type PluginDecoratorsOptions = {
  version: DecoratorsVersion;
  decoratorsBeforeExport?: boolean;
};

export type BasePresetOptions = {
  presetEnv?: PresetEnvOptions | false;
  presetTypeScript?: Record<string, unknown> | false;
  pluginDecorators?: PluginDecoratorsOptions | false;
};

export type WebPresetOptions = BasePresetOptions & {
  presetEnv: PresetEnvOptions & {
    targets: PresetEnvTargets;
    useBuiltIns: PresetEnvBuiltIns;
  };
  pluginTransformRuntime?: Record<string, unknown> | false;
};

export type NodePresetOptions = BasePresetOptions & {
  // empty
};
