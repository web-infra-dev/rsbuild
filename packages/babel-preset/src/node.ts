import { deepmerge } from '@rsbuild/shared/deepmerge';
import { generateBaseConfig } from './base';
import type { BabelConfig, NodePresetOptions } from './types';

export const getBabelConfigForNode = (options: NodePresetOptions = {}) => {
  const mergedOptions = deepmerge(
    {
      presetEnv: {
        targets: ['node >= 14'],
      },
    },
    options,
  );

  const config = generateBaseConfig(mergedOptions);

  config.plugins?.push(require.resolve('babel-plugin-dynamic-import-node'));

  return config;
};

export default function (
  api: { cache: (flag: boolean) => void },
  options: NodePresetOptions = {},
): BabelConfig {
  api.cache(true);
  return getBabelConfigForNode(options);
}
