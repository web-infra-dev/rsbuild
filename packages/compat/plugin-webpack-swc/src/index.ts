export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
