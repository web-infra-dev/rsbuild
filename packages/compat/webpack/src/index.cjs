/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
