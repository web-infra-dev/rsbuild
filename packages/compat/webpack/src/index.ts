export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
