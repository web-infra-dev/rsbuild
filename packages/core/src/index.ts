/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
