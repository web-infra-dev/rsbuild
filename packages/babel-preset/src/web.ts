import { join } from 'node:path';
import { readFileSync } from 'node:fs';
import { generateBaseConfig } from './base';
import type { BabelConfig, WebPresetOptions } from './types';
import type { PresetEnvOptions } from '@rsbuild/plugin-babel';

const getCoreJsVersion = (corejsPkgPath: string) => {
  try {
    const { version } = JSON.parse(readFileSync(corejsPkgPath, 'utf-8'));
    const [major, minor] = version.split('.');
    return `${major}.${minor}`;
  } catch (err) {
    return '3';
  }
};

const getDefaultPresetEnvOption = (
  options: WebPresetOptions,
): PresetEnvOptions | false => {
  if (options.presetEnv === false) {
    return false;
  }

  return {
    bugfixes: true,
    // core-js is required for web target
    corejs: options.presetEnv?.useBuiltIns
      ? {
          version: getCoreJsVersion(require.resolve('core-js/package.json')),
          proposals: true,
        }
      : undefined,
  };
};

export const getBabelConfigForWeb = (options: WebPresetOptions) => {
  if (options.presetEnv !== false) {
    options.presetEnv = {
      ...getDefaultPresetEnvOption(options),
      ...options.presetEnv,
    };
  }

  const config = generateBaseConfig(options);
  const { pluginTransformRuntime = {} } = options;

  // Skip plugin-transform-runtime when testing
  if (pluginTransformRuntime) {
    config.plugins?.push([
      require.resolve('@babel/plugin-transform-runtime'),
      {
        version: require('@babel/runtime/package.json').version,
        // this option has been deprecated
        // but enabling it can help to reduce bundle size
        useESModules: true,
        ...pluginTransformRuntime,
      },
    ]);
  }

  config.plugins?.push(join(__dirname, './pluginLockCorejsVersion'));

  return config;
};

export default function (
  api: { cache: (flag: boolean) => void },
  options: WebPresetOptions,
): BabelConfig {
  api.cache(true);
  return getBabelConfigForWeb(options);
}
