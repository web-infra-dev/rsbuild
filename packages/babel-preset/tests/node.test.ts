import { getBabelConfigForNode } from '../src/node';

test('should provide node preset as expected', () => {
  expect(getBabelConfigForNode()).toMatchSnapshot();
});

test('should allow to override target node version', () => {
  expect(
    getBabelConfigForNode({
      presetEnv: {
        targets: ['node >= 20'],
      },
    }),
  ).toMatchSnapshot();
});
