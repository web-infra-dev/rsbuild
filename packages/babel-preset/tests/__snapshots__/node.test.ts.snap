// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow to override target node version 1`] = `
{
  "plugins": [
    "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
    "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
      {
        "proposal": "minimal",
      },
    ],
    "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-dynamic-import-node/lib/index.js",
  ],
  "presets": [
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
      {
        "exclude": [
          "transform-typeof-symbol",
        ],
        "modules": "commonjs",
        "targets": [
          "node >= 20",
        ],
      },
    ],
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
      {
        "allExtensions": true,
        "allowDeclareFields": true,
        "allowNamespaces": true,
        "isTSX": true,
        "optimizeConstEnums": true,
      },
    ],
  ],
}
`;

exports[`should provide node preset as expected 1`] = `
{
  "plugins": [
    "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
    "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
      {
        "proposal": "minimal",
      },
    ],
    "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-dynamic-import-node/lib/index.js",
  ],
  "presets": [
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
      {
        "exclude": [
          "transform-typeof-symbol",
        ],
        "modules": "commonjs",
        "targets": [
          "node >= 14",
        ],
      },
    ],
    [
      "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
      {
        "allExtensions": true,
        "allowDeclareFields": true,
        "allowNamespaces": true,
        "isTSX": true,
        "optimizeConstEnums": true,
      },
    ],
  ],
}
`;
